Add todo widget

kamilwawrzyczek opened the pull request:
## Summary
Added todo list widget. Widget can redirect to full list view when clicked on header and it also can mark items as completed when clicked on checkbox. Resolves #2374.

## Screenshots
![todo_list_config](https://github.com/user-attachments/assets/ca6ed974-edc2-4825-bc6b-d205c5f4a4cb)
![todo_list_dark](https://github.com/user-attachments/assets/5891f2a9-8f6c-4b97-a059-912509e46df7)
![todo_list_white](https://github.com/user-attachments/assets/7788b778-589f-433f-8806-c2a79f454239)


## Link to pull request in Documentation repository
Documentation: [home-assistant/companion.home-assistant#](https://github.com/home-assistant/companion.home-assistant/pull/1188)

## Any other notes
- added remote views compact library as it simplified widget with list creation
- why trailing commas in kotlin are not allowed? I needed to fix up all of my files 😅

kamilwawrzyczek commented on app/src/main/kotlin/io/homeassistant/companion/android/widgets/todo/TodoWidgetConfigureActivity.kt:
```diff
@@ -0,0 +1,204 @@
+    private fun onSetupWidget() {
+        if (intent.extras?.getBoolean(ManageWidgetsViewModel.CONFIGURE_REQUEST_LAUNCHER, false) == true) {
+            if (
+                Build.VERSION.SDK_INT >= Build.VERSION_CODES.O &&
```
min for app is 21, only automotive have 29 and widget is not supported there

kamilwawrzyczek commented on app/src/main/kotlin/io/homeassistant/companion/android/widgets/todo/TodoWidgetConfigureActivity.kt:
```diff
@@ -0,0 +1,204 @@
+        }
+    }
+
+    @RequiresApi(Build.VERSION_CODES.O)
```
min for app is 21, only automotive have 29 and widget is not supported there

kamilwawrzyczek replied:
other widgets are done the same way, it do not matter as it will be just ignored on older versions

kamilwawrzyczek commented on app/src/main/res/layout/widget_todo.xml:
```diff
@@ -0,0 +1,50 @@
+            android:padding="8dp"
+            android:textColor="?colorWidgetOnBackground"
+            android:textSize="16sp"
+            android:drawableEnd="@drawable/ic_launch" />
```
it do not work with RemoteViews, it need to be drawableEnd

kamilwawrzyczek commented on app/src/main/res/layout/widget_todo.xml:
```diff
@@ -0,0 +1,50 @@
+            android:padding="8dp"
+            android:textColor="?colorWidgetOnBackground"
+            android:textSize="16sp"
+            android:drawableEnd="@drawable/ic_launch" />
```
it do not work with RemoteViews, it need to be drawableEnd

kamilwawrzyczek commented on app/src/main/res/xml/todo_widget_info.xml:
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="utf-8"?>
+<appwidget-provider xmlns:android="http://schemas.android.com/apk/res/android"
+    android:configure="io.homeassistant.companion.android.widgets.todo.TodoWidgetConfigureActivity"
+    android:initialLayout="@layout/widget_todo"
+    android:minWidth="250dp"
+    android:minHeight="100dp"
+    android:minResizeWidth="150dp"
+    android:minResizeHeight="80dp"
+    android:resizeMode="vertical|horizontal"
+    android:targetCellWidth="4"
+    android:targetCellHeight="3"
+    android:updatePeriodMillis="1800000"
+    android:widgetFeatures="reconfigurable"
+    android:description="@string/todo_widget_desc"
+    android:widgetCategory="home_screen"
+    android:previewImage="@drawable/widget_example_todo"
+    android:previewLayout="@layout/widget_todo_wrapper_dynamiccolor" />
```
it is in the manifest

kamilwawrzyczek commented on app/src/main/res/xml/todo_widget_info.xml:
```diff
@@ -0,0 +1,17 @@
+    android:minResizeWidth="150dp"
+    android:minResizeHeight="80dp"
+    android:resizeMode="vertical|horizontal"
+    android:targetCellWidth="4"
```
other widgets are done the same way, it do not matter as it will be just ignored on older versions

kamilwawrzyczek commented on app/src/main/res/xml/todo_widget_info.xml:
```diff
@@ -0,0 +1,17 @@
+    android:minResizeWidth="150dp"
+    android:minResizeHeight="80dp"
+    android:resizeMode="vertical|horizontal"
+    android:targetCellWidth="4"
```
other widgets are done the same way, it do not matter as it will be just ignored on older versions

kamilwawrzyczek commented on app/src/main/res/xml/todo_widget_info.xml:
```diff
@@ -0,0 +1,17 @@
+    android:minResizeHeight="80dp"
+    android:resizeMode="vertical|horizontal"
+    android:targetCellWidth="4"
+    android:targetCellHeight="3"
```
other widgets are done the same way, it do not matter as it will be just ignored on older versions

kamilwawrzyczek commented on app/src/main/res/xml/todo_widget_info.xml:
```diff
@@ -0,0 +1,17 @@
+    android:minResizeHeight="80dp"
+    android:resizeMode="vertical|horizontal"
+    android:targetCellWidth="4"
+    android:targetCellHeight="3"
```
other widgets are done the same way, it do not matter as it will be just ignored on older versions

kamilwawrzyczek commented on app/src/main/res/xml/todo_widget_info.xml:
```diff
@@ -0,0 +1,17 @@
+    android:targetCellWidth="4"
+    android:targetCellHeight="3"
+    android:updatePeriodMillis="1800000"
+    android:widgetFeatures="reconfigurable"
```
other widgets are done the same way, it do not matter as it will be just ignored on older versions

kamilwawrzyczek commented on app/src/main/res/xml/todo_widget_info.xml:
```diff
@@ -0,0 +1,17 @@
+    android:description="@string/todo_widget_desc"
+    android:widgetCategory="home_screen"
+    android:previewImage="@drawable/widget_example_todo"
+    android:previewLayout="@layout/widget_todo_wrapper_dynamiccolor" />
```
other widgets are done the same way, it do not matter as it will be just ignored on older versions

kamilwawrzyczek commented on app/src/main/res/xml/todo_widget_info.xml:
```diff
@@ -0,0 +1,17 @@
+    android:description="@string/todo_widget_desc"
+    android:widgetCategory="home_screen"
+    android:previewImage="@drawable/widget_example_todo"
+    android:previewLayout="@layout/widget_todo_wrapper_dynamiccolor" />
```
other widgets are done the same way, it do not matter as it will be just ignored on older versions

TimoPtr commented:
![image](https://github.com/user-attachments/assets/00f749ad-2251-48fc-83d6-322d8e654640)
Could you add some items in the sample of the widget?

TimoPtr commented:
Few UX remarks
![image](https://github.com/user-attachments/assets/9f5b0c81-12ba-4583-8ca9-d0c54a48f4c9)
- **Empty List Hint:** If the list is empty, I would expect a hint or message indicating that there are no items. This helps users understand that the list is intentionally empty and not that content failed to load.
- **Quick Add Button:** A quick add button to add an entry directly from the list would be a nice enhancement. Similar to the Google Calendar widget, this would allow users to add items quicker.
- **Simplify Access to the List:** The open button that currently opens the list could be removed. Instead, we could make the widget's title clickable to open the list, which would streamline the user interface. 

https://github.com/user-attachments/assets/79224ffb-89a0-4c77-a595-401d3be68b47


- **Loading Animation Doesn't Block Background Clicks:** I noticed that while the loading animation is active, it doesn't prevent clicks on the view behind it. It would be beneficial to block interactions with the background during loading to prevent unintended actions.
- **Clickable Areas on List Items:** Clicking on an entire item line triggers the click animation but doesn't perform an action unless it is the checkbox that is clicked. It might improve usability to make only the checkbox clickable or to add functionality when the whole line is clicked.
- **Consistency in Loading Animation Styling:** The loading animation's background has sharp corners, which doesn't match the widget's rounded corners. Adjusting it to match the widget's style would enhance visual consistency.
![image](https://github.com/user-attachments/assets/c3697ebe-2746-42b0-b787-e65b9f76ae54)
- **Error Handling on Item Click:** When an error occurs after clicking on an item, the loading view isn't dismissed, and no error message is shown. Implementing proper error handling to dismiss the loading view and provide feedback would improve the user experience.

bgoncal commented:
Will the data always reflect realtime information? Otherwise I would replace the top right icon by a refresh button and if the user wants to open the list inside the App it can just tap the header.

dshokouhi commented on common/src/main/res/values/strings.xml:
```diff
@@ -252,6 +252,7 @@
     <string name="entity_id">Entity ID</string>
     <string name="entity_id_name">Entity ID: %1$s</string>
     <string name="entity_widget_desc">View any entity\'s current state and toggle it</string>
+    <string name="todo_widget_desc">View configured todo list</string>
```
```suggestion
    <string name="todo_widget_desc">View configured to-do list</string>
```

Lets match the frontend and HA documentation here to keep thing aligned, can you please also update the other strings to add the missing hyphen please?

https://www.home-assistant.io/integrations/todo/

![image](https://github.com/user-attachments/assets/2cbbc99c-6727-4c04-bbb0-456ad029f9f3)

kamilwawrzyczek replied:
done

kamilwawrzyczek commented:
I believe all comments should be sorted:
- added empty state
- fixed non-image preview 
- add blocking click with loading + loading styling
- disable ripple on list
- force trigger update to fix loading getting stuck

Widget should update on its own, but I also added refresh button just in case it is needed.
I added also add button but it redirects to the todo page on fronted. I did not found an easy way to trigger add dialog. Any ideas here? Only thing that comes to my mind is to create native activity to manage todos but it feel like overkill to have this activity just to support adding from the widget.

@TimoPtr @bgoncal

bgoncal commented:
Can we make "show completed" optional? And about the "add item", I think opening the list view inside the app is already enough since the UI already displays a text field to add a new item anyways 
![CleanShot 2025-04-01 at 09 58 41@2x](https://github.com/user-attachments/assets/37e83e3d-dbc8-450e-82d8-1eac5eaf013f)

kamilwawrzyczek commented:
Show completed option added to settings

bgoncal commented:
Usually those floating action buttons are positioned on the bottom right, but I am an iOS user so I'll leave to @TimoPtr to decide, in general it looks fine by me.

dshokouhi commented on common/src/main/kotlin/io/homeassistant/companion/android/database/AppDatabase.kt:
```diff
@@ -100,7 +103,7 @@ import timber.log.Timber
         Setting::class
     ],
-    version = 49,
+    version = 51,
```
We should only have 1 migration in the PR

kamilwawrzyczek replied:
rolled-back to version 50

jpelgrom commented:
> Usually those floating action buttons are positioned on the bottom right, but I am an iOS user so I'll leave to @TimoPtr to decide, in general it looks fine by me.

Android follows Material Design so I'd also say it should be in the bottom right. This looks odd to me, also because of the size.

kamilwawrzyczek commented:
Are you referring to the plus button on header? Those are not floating action button just regular one, google calendar widget have similar one on the top. Also keeping it next to header is saving space as if I move it to bottom they will be on its own line where right now we can show 1-2 to-dos instead.
I could do smaller size but it might be hard to hit and I believe the one on google calendar is even bigger than this one :)

jpelgrom commented:
I was thinking of the Google Keep 'Single note' widget as an example which is a shopping list and has a FAB in the bottom right :) Because it is a perfect circle without a shadow your current one doesn't look right to me, the example you mentioned is M3 so less round and also doesn't clash with the corner radius.

kamilwawrzyczek commented:
Keep have 1 button and it is actually a fab (is over the content) I would need to do 2 buttons. IMO it would be less readable.

I prepared version with smaller buttons, also the button radius depends on style e.g dynamic color have round button (I used existing styles).
![examples](https://github.com/user-attachments/assets/8a15f42d-d793-4688-8748-776b2bb920c3)

dshokouhi commented:
Personally am not sure a refresh button is needed here if a websocket is used to keep the list up to date. We have a template widget which also subscribes to a websocket and will shows a cloud icon when not connected to the server. That should be part of the standard widget like the entity state widget which also uses a subscription. The subscription is also tied to screen on/off events so it should be relatively up to date assuming the server is reachable.

kamilwawrzyczek commented:
I actually did use that during my testing. Not really sure what did happen because I did not have my phone connected but the list stopped updating (and I am using this BaseWidgetProvider that have all of those logic to subscribe for updates). Refresh did fix that. Other widgets also have refresh functionality so I think it would be safer to leave it and remove later if it is not needed.

dshokouhi commented on common/src/main/res/values/strings.xml:
```diff
@@ -1045,6 +1047,16 @@
+    <string name="widget_todo_add">Add to-do</string>
+    <string name="widget_todo_refresh">Refresh to-do</string>
+    <string name="widget_todo_label">To-do list</string>
+    <string name="widget_todo_empty">No items in the to-do list</string>
```
```suggestion
    <string name="widget_todo_empty">You have no to-do items!</string>
```

To match the frontend english string

![image](https://github.com/user-attachments/assets/257d4fcd-eefd-4d39-bd9d-5f5824a06fa9)

kamilwawrzyczek replied:
updated

dshokouhi commented:
Can we make the preview example smaller theres a lot of unused space in the current preview

![image](https://github.com/user-attachments/assets/a5e463d5-3b3c-4d85-8e69-404a934d2deb)

Everytime I drag the widget to my home screen I get a toast error message, doesnt happen when I drag the template widget for example

![image](https://github.com/user-attachments/assets/b5fe7356-6b9b-4fe5-99b5-086363e7c347)

the error that shows up in the logs is

```
2025-04-01 11:19:32.983 29657-31183 WebSocketRepositoryImpl io....stant.companion.android.debug  E  Websocket: onFailure (Ask Gemini)
                                                                                                    java.net.SocketException: Software caused connection abort
                                                                                                    	at java.net.SocketInputStream.socketRead0(Native Method)
                                                                                                    	at java.net.SocketInputStream.socketRead(SocketInputStream.java:118)
                                                                                                    	at java.net.SocketInputStream.read(SocketInputStream.java:173)
                                                                                                    	at java.net.SocketInputStream.read(SocketInputStream.java:143)
                                                                                                    	at com.android.org.conscrypt.ConscryptEngineSocket$SSLInputStream.readFromSocket(ConscryptEngineSocket.java:989)
                                                                                                    	at com.android.org.conscrypt.ConscryptEngineSocket$SSLInputStream.processDataFromSocket(ConscryptEngineSocket.java:953)
                                                                                                    	at com.android.org.conscrypt.ConscryptEngineSocket$SSLInputStream.readUntilDataAvailable(ConscryptEngineSocket.java:868)
                                                                                                    	at com.android.org.conscrypt.ConscryptEngineSocket$SSLInputStream.read(ConscryptEngineSocket.java:841)
                                                                                                    	at okio.InputStreamSource.read(JvmOkio.kt:93)
                                                                                                    	at okio.AsyncTimeout$source$1.read(AsyncTimeout.kt:153)
                                                                                                    	at okio.RealBufferedSource.request(RealBufferedSource.kt:211)
                                                                                                    	at okio.RealBufferedSource.require(RealBufferedSource.kt:204)
                                                                                                    	at okio.RealBufferedSource.readByte(RealBufferedSource.kt:214)
                                                                                                    	at okhttp3.internal.ws.WebSocketReader.readHeader(WebSocketReader.kt:124)
                                                                                                    	at okhttp3.internal.ws.WebSocketReader.processNextFrame(WebSocketReader.kt:107)
                                                                                                    	at okhttp3.internal.ws.RealWebSocket.loopReader(RealWebSocket.kt:307)
                                                                                                    	at okhttp3.internal.ws.RealWebSocket$connect$1.onResponse(RealWebSocket.kt:195)
                                                                                                    	at okhttp3.internal.connection.RealCall$AsyncCall.run(RealCall.kt:529)
                                                                                                    	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
                                                                                                    	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:644)
                                                                                                    	at java.lang.Thread.run(Thread.java:1012)
```

TimoPtr commented:
I got the issue too about the link to the WebSocket is broken and stop refreshing the list. @dshokouhi did you ever see the issue in the template widget?

dshokouhi commented:
> @dshokouhi did you ever see the issue in the template widget?

no i have not seen the issue before....was very easily able to reproduce the issue with the to-do widget but same steps are not replicable on existing widgets like template

kamilwawrzyczek commented:
When widget is added system is calling update widget to get a remoteview and at this point there is nothing in the DAO. I removed the toast there as it is expected that dao is empty for that case.
As for the stacktrace it is not related - I think it is just websocket getting killed when app is in background and after app restart it just getting the callback for this (it is possible to reproduce it even without my changes).
Preview should also be smaller now.

TimoPtr commented:
I did end up into a situation where the loading view was always display and I was not able to interact with the widget anymore

Just clicked on an item after like 10mn of inactivity.

```
2025-04-03 09:23:23.921  5795-7313  WebSocketRepositoryImpl io....stant.companion.android.debug  E  Websocket: onFailure (Ask Gemini)
java.net.SocketException: Software caused connection abort
	at java.net.SocketInputStream.socketRead0(Native Method)
	at java.net.SocketInputStream.socketRead(SocketInputStream.java:118)
	at java.net.SocketInputStream.read(SocketInputStream.java:173)
	at java.net.SocketInputStream.read(SocketInputStream.java:143)
	at okio.InputStreamSource.read(JvmOkio.kt:93)
	at okio.AsyncTimeout$source$1.read(AsyncTimeout.kt:153)
	at okio.RealBufferedSource.request(RealBufferedSource.kt:211)
	at okio.RealBufferedSource.require(RealBufferedSource.kt:204)
	at okio.RealBufferedSource.readByte(RealBufferedSource.kt:214)
	at okhttp3.internal.ws.WebSocketReader.readHeader(WebSocketReader.kt:124)
	at okhttp3.internal.ws.WebSocketReader.processNextFrame(WebSocketReader.kt:107)
	at okhttp3.internal.ws.RealWebSocket.loopReader(RealWebSocket.kt:307)
	at okhttp3.internal.ws.RealWebSocket$connect$1.onResponse(RealWebSocket.kt:195)
	at okhttp3.internal.connection.RealCall$AsyncCall.run(RealCall.kt:529)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:644)
	at java.lang.Thread.run(Thread.java:1012)
2025-04-03 09:23:52.491  5795-6039  ServerConnectionInfo    io....stant.companion.android.debug  D  Using external URL
2025-04-03 09:23:52.505  5795-6039  ServerConnectionInfo    io....stant.companion.android.debug  D  Using external URL
2025-04-03 09:23:52.507  5795-7313  okhttp.OkHttpClient     io....stant.companion.android.debug  I  --> GET http://192.168.11.15:8123/api/websocket
2025-04-03 09:23:52.508  5795-7313  okhttp.OkHttpClient     io....stant.companion.android.debug  I  User-Agent: Home Assistant/2025.3.5-beta.0.29+20250403T070634Z-1 (Android 15; Pixel 6)
2025-04-03 09:23:52.508  5795-7313  okhttp.OkHttpClient     io....stant.companion.android.debug  I  Upgrade: websocket
2025-04-03 09:23:52.508  5795-7313  okhttp.OkHttpClient     io....stant.companion.android.debug  I  Connection: Upgrade
2025-04-03 09:23:52.508  5795-7313  okhttp.OkHttpClient     io....stant.companion.android.debug  I  Sec-WebSocket-Key: XlV+6ug+5DiO/F+jr+WHsw==
2025-04-03 09:23:52.508  5795-7313  okhttp.OkHttpClient     io....stant.companion.android.debug  I  Sec-WebSocket-Version: 13
2025-04-03 09:23:52.508  5795-7313  okhttp.OkHttpClient     io....stant.companion.android.debug  I  Sec-WebSocket-Extensions: permessage-deflate
2025-04-03 09:23:52.508  5795-7313  okhttp.OkHttpClient     io....stant.companion.android.debug  I  --> END GET
2025-04-03 09:24:02.535  5795-7313  okhttp.OkHttpClient     io....stant.companion.android.debug  I  <-- HTTP FAILED: java.net.SocketTimeoutException: failed to connect to /192.168.11.15 (port 8123) from /192.168.11.3 (port 41176) after 10000ms
2025-04-03 09:24:02.539  5795-7313  WebSocketRepositoryImpl io....stant.companion.android.debug  E  Websocket: onFailure (Ask Gemini)
java.net.SocketTimeoutException: failed to connect to /192.168.11.15 (port 8123) from /192.168.11.3 (port 41176) after 10000ms
	at libcore.io.IoBridge.connectErrno(IoBridge.java:235)
	at libcore.io.IoBridge.connect(IoBridge.java:179)
	at java.net.PlainSocketImpl.socketConnect(PlainSocketImpl.java:142)
	at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:390)
	at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:230)
	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:212)
	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:436)
	at java.net.Socket.connect(Socket.java:646)
	at okhttp3.internal.platform.Platform.connectSocket(Platform.kt:145)
	at okhttp3.internal.connection.ConnectPlan.connectSocket(ConnectPlan.kt:271)
	at okhttp3.internal.connection.ConnectPlan.connectTcp(ConnectPlan.kt:139)
	at okhttp3.internal.connection.FastFallbackExchangeFinder$launchTcpConnect$1.runOnce(FastFallbackExchangeFinder.kt:141)
	at okhttp3.internal.concurrent.TaskRunner.runTask(TaskRunner.kt:124)
	at okhttp3.internal.concurrent.TaskRunner.access$runTask(TaskRunner.kt:44)
	at okhttp3.internal.concurrent.TaskRunner$runnable$1.run(TaskRunner.kt:73)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:644)
	at java.lang.Thread.run(Thread.java:1012)
```

I killed the app open it again it didn't do anything. The only thing that worked was to lock my screen and unlock it. I guess it reloaded the widget.

TimoPtr commented on app/src/main/kotlin/io/homeassistant/companion/android/widgets/todo/TodoWidgetViewModel.kt:
```diff
@@ -0,0 +1,128 @@
+import kotlinx.coroutines.launch
+
+@HiltViewModel
+class TodoWidgetViewModel @Inject constructor(
+    private val application: Application,
+    private val todoWidgetDao: TodoWidgetDao,
+    private val serverManager: ServerManager
+) : AndroidViewModel(application) {
```
I think you could remove the need of Application here. I find it safer to not manipulate a context even if it is the application context. For the color you can simply send the ID in the Intent and the Widget will retrieve the color. For the component you can still set it in the Activity instead of the viewModel.

TimoPtr replied:
After re-reading this, I'm actually wondering about the color stored inside the Entry. If we change the color at some points of `colorWidgetButtonLabelBlack` it won't reflect in the entry since we store the value of the color not the ID. It will be updated when someone update the Entry.

Storing the ID of the color in the Entry might not be a good idea if we remove the color one day it would crash since it won't find the color (What I propose in the above comments 👎🏻).

Do you really want to keep the color inside the Entry? Can't we get it from a theme?

kamilwawrzyczek replied:
it is done the same for e.g entity widget and it is used only for transparent theme to select white/black color of the text depending on user background. I think it is fine to keep resolved color there. If user want to change the widget they can do reconfigure but if the don't then lets keep it as it is.

TimoPtr commented on app/src/main/kotlin/io/homeassistant/companion/android/widgets/todo/TodoWidgetViewModel.kt:
```diff
@@ -0,0 +1,128 @@
+
+    val entities: StateFlow<List<Entity<*>>> = snapshotFlow { selectedServerId }
+        .distinctUntilChanged()
+        .mapLatest { serverId ->
```
I think mapLatest needs `ExperimentalCoroutinesApi`

TimoPtr commented on app/src/main/kotlin/io/homeassistant/companion/android/widgets/todo/TodoWidgetViewModel.kt:
```diff
@@ -0,0 +1,128 @@
+    }
+
+    fun prepareData(): Intent? {
+        if (!isValidSelection()) return null
```
If one it happens we won't know why for instance here.

TimoPtr commented on common/src/main/kotlin/io/homeassistant/companion/android/common/data/websocket/impl/entities/GetTodosResponse.kt:
```diff
@@ -0,0 +1,21 @@
+    val response: Map<String, TodoResponse>
+) {
+
+    @JsonIgnoreProperties(ignoreUnknown = true)
+    data class TodoResponse(
+        val items: List<Todo>
+    )
+
+    @JsonIgnoreProperties(ignoreUnknown = true)
+    data class Todo(
+        val uid: String,
+        val summary: String,
+        val status: String
+    )
```
https://developers.home-assistant.io/docs/core/entity/todo/#todoitem I think we should support the case null no?

kamilwawrzyczek replied:
If I understand this correctly then it should never be null when it is added, but I am all for playing it safe - will add nullability there.

TimoPtr commented on common/src/main/kotlin/io/homeassistant/companion/android/database/AppDatabase.kt:
```diff
@@ -125,7 +128,8 @@ import timber.log.Timber
         AutoMigration(from = 47, to = 48),
-        AutoMigration(from = 48, to = 49)
+        AutoMigration(from = 48, to = 49),
+        AutoMigration(from = 49, to = 50)
```
```suggestion
        AutoMigration(from = 49, to = 50),
```
This has been allowed in a recent PR. It will make future changes better in Git to have a comma here.

TimoPtr commented on common/src/main/kotlin/io/homeassistant/companion/android/database/widget/TodoWidgetEntity.kt:
```diff
@@ -0,0 +1,21 @@
+data class TodoWidgetEntity(
+    @PrimaryKey
+    override val id: Int,
+    @ColumnInfo(name = "server_id", defaultValue = "0")
```
Any reason to have a defaultValue here ?

kamilwawrzyczek replied:
copy pasted from other widgets

TimoPtr reviewed (commented):
All this comments are mostly to improve the readability and maintenance but overall it fits well in the existing widgets 👍🏻

kamilwawrzyczek commented on app/src/main/kotlin/io/homeassistant/companion/android/widgets/todo/TodoWidgetConfigureActivity.kt:
```diff
@@ -0,0 +1,229 @@
+    @RequiresApi(Build.VERSION_CODES.O)
+    private fun requestPinWidget() {
+        getSystemService<AppWidgetManager>()?.requestPinAppWidget(
+            ComponentName(this, MediaPlayerControlsWidget::class.java),
```
I forgot to change it after copy the code 😅

dshokouhi commented on common/src/main/res/values/strings.xml:
```diff
@@ -252,6 +252,7 @@
     <string name="entity_id">Entity ID</string>
     <string name="entity_id_name">Entity ID: %1$s</string>
     <string name="entity_widget_desc">View any entity\'s current state and toggle it</string>
+    <string name="todo_widget_desc">View configured to-do list</string>
```
```suggestion
    <string name="todo_widget_desc">View any configured to-do list</string>
```

The other widgets also describe they have actions we could also mention that here such as marking items as completed, my suggestion was just for readability and hopefully help translators :)

dshokouhi commented on app/src/main/kotlin/io/homeassistant/companion/android/widgets/todo/TodoWidgetConfigureActivity.kt:
```diff
@@ -0,0 +1,229 @@
+                modifier = Modifier.fillMaxWidth(),
+                onClick = { onAddWidget() }
+            ) {
+                Text(stringResource(R.string.add_widget))
```
This should update to "Update Widget" if the user is editing an existing widget

kamilwawrzyczek commented:
I hope I did not miss anything and resolved all of the comments. If there is something missing let me know.
I also fixed this loading issue (I was able to reproduce it by clicking complete todo when there was not internet connection).

For next few days I won't have access to my computer so I will resolve any potential issues next week.

TimoPtr commented on app/src/main/kotlin/io/homeassistant/companion/android/widgets/todo/TodoWidget.kt:
```diff
@@ -0,0 +1,410 @@
+        }
+        setOnClickPendingIntent(
+            R.id.widget_todo_title,
+            PendingIntent.getBroadcast(
+                context,
+                entity.id,
+                openIntent,
+                PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE,
+            ),
```
This one is also duplicated

TimoPtr commented on app/src/main/kotlin/io/homeassistant/companion/android/widgets/todo/TodoWidget.kt:
```diff
@@ -0,0 +1,410 @@
+            ),
+        )
+
+        val toggleIntent = Intent(context, TodoWidget::class.java).apply {
+            action = TOGGLE_TODO_ITEM
+            putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, entity.id)
+            putExtra(EXTRA_SERVER_ID, entity.serverId)
+            putExtra(EXTRA_ENTITY_ID, entity.entityId)
+        }
```
You could create a small function that create the Intent that only takes the action as parameter since it's repeated twice.

TimoPtr commented on app/src/main/kotlin/io/homeassistant/companion/android/widgets/todo/TodoWidget.kt:
```diff
@@ -0,0 +1,410 @@
+    override fun saveEntityConfiguration(context: Context, extras: Bundle?, appWidgetId: Int) {
+        if (extras == null) return
+        if (!extras.containsKey(EXTRA_SERVER_ID) || !extras.containsKey(EXTRA_ENTITY_ID)) {
+            return
```
Add a log saying missing extra or keys

TimoPtr commented on app/src/main/kotlin/io/homeassistant/companion/android/widgets/todo/TodoWidgetViewModel.kt:
```diff
@@ -0,0 +1,134 @@
+    }
+
+    fun prepareData(componentName: ComponentName): Intent? {
+        if (!isValidSelection()) {
+            Timber.d("Widget data is invalid")
+            return null
+        }
+        if (widgetId == AppWidgetManager.INVALID_APPWIDGET_ID) {
+            Timber.d("Widget ID is invalid")
+            return null
+        }
```
could be a warning

TimoPtr commented on app/src/main/kotlin/io/homeassistant/companion/android/widgets/todo/TodoWidgetViewModel.kt:
```diff
@@ -0,0 +1,134 @@
+                .orEmpty()
+                .filter { entity -> entity.domain == "todo" }
+        }.stateIn(viewModelScope, SharingStarted.WhileSubscribed(500), emptyList())
+    var selectedEntityId by mutableStateOf<String?>(null)
+        private set
+
+    var selectedBackgroundType by mutableStateOf(WidgetBackgroundType.DAYNIGHT)
+        private set
+    var textColorIndex by mutableIntStateOf(0)
+        private set
+    var showCompletedState by mutableStateOf(true)
+        private set
+    var isUpdateWidget by mutableStateOf(false)
+        private set
```
Exposing the setter is ok IMO since you are using the `by` operators it does not expose the mutableState directly but the underlying type.

It would allow you to remove all the manual setters.

kamilwawrzyczek replied:
I did that on purpose to make sure that VM is only entity allowed to update state. If I expose set then potentially UI can update state wrongly or multiple places could update this state (this makes app hard to maintain if we do not have single path for updates). This solution allows me to easily add some additional logic on state change if I wanted to e.g updating server should clear entityId.

TimoPtr commented:
While using the widget on my phone, I saw the disconnect icon on the entity widget when the WebSocket is down. Could you integrate it also in your widget it is quite helpful.

If you can't we should create an issue and someone will do it in another PR.

kamilwawrzyczek commented:
![Screenshot_20250413_105903](https://github.com/user-attachments/assets/88a469ab-b093-4b15-9f90-9c2112d1f75f) error indicator added.

TimoPtr commented on app/src/main/kotlin/io/homeassistant/companion/android/widgets/todo/TodoWidget.kt:
```diff
@@ -222,9 +225,55 @@ class TodoWidget : BaseWidgetProvider() {
+            updateLatestData(entity.id, name, todos)
+            TodoWidgetData(
+                name = name,
+                itemsByStatus = todos.groupBy { it.isComplete },
```
You can create a small function that you can reuse or even handle this in a secondary ctor of `TodoWidgetData`

TimoPtr reviewed (commented):
Thanks for taking all my requests in consideration. Let's wait for @dshokouhi review before getting this one merged. 
I look very nice 👏🏻

dshokouhi reviewed (approved):
Ok was able to spend sometime testing functionally things look great and update as expected for the most part. This will make a lot of our users happy.

There are only 2 case where the widget does not instant update and that is directly related to the websocket subscription updates all widgets use. Only when a To-Do item is renamed or re-ordered does the subscription not update. Screen on events capture this change. Looking at the HA core code there is another websocket subscription we appear to need for these 2 cases. @kamilwawrzyczek  can you please [create an enhancement](https://github.com/home-assistant/android/issues/new?template=feature_request.md) for the app for a future PR to add support for it please? Lets not increase scope here. 

https://github.com/home-assistant/core/blob/dev/homeassistant/components/todo/__init__.py#L322
